src/libraries/System.Text.Json/src/System/Text/Json/JsonHelpers.cs

```diff
@@ -355,48 +355,55 @@ public static bool AreEqualJsonNumbers(ReadOnlySpan<byte> left, ReadOnlySpan<byt
                 return false;
             }
 
-            if (leftIntegral.Length == rightIntegral.Length)
+            // Need to check that the concatenated integral and fractional parts are equal;
+            // break each representation into three parts such that their lengths exactly match.
+            ReadOnlySpan<byte> leftFirst;
+            ReadOnlySpan<byte> leftMiddle;
+            ReadOnlySpan<byte> leftLast;
+
+            ReadOnlySpan<byte> rightFirst;
+            ReadOnlySpan<byte> rightMiddle;
+            ReadOnlySpan<byte> rightLast;
+
+            int diff = leftIntegral.Length - rightIntegral.Length;
+            switch (diff)
             {
-                return leftIntegral.SequenceEqual(rightIntegral) &&
-                    leftFractional.SequenceEqual(rightFractional);
+                case < 0:
+                    leftFirst = leftIntegral;
+                    leftMiddle = leftFractional.Slice(0, -diff);
+                    leftLast = leftFractional.Slice(-diff);
+                    int rightOffset = rightIntegral.Length + diff;
+                    rightFirst = rightIntegral.Slice(0, rightOffset);
+                    rightMiddle = rightIntegral.Slice(rightOffset);
+                    rightLast = rightFractional;
```

Comment:
nit: It's very easy to misread or misinterpret the logic here since `diff` is negative.

It would be beneficial if you started the handling by negating `diff` (getting its absolute value) and then using subtraction/etc instead

Reply:
I figured that the `case < 0:` clause above should self-document this.